File: tools/print_config_exhaustive/lint.js
// @ts-check
/// <reference types="node" />

/**
 * This file is a linter that verifies that the `printConfig` function
 * references every field in the `NormalConfigData` struct.
 */

import fs from "node:fs/promises";

const NORMAL_CONFIG_DATA_PATH = "internal/config/configdomain/normal_config_data.go";
const PRINT_CONFIG_PATH = "internal/cmd/config/root.go";
const ACCEPTABLE_MISSING_FIELDS = [
  "Aliases",
  "BranchTypeOverrides",
];

const normalConfigDataFile = await fs.readFile(NORMAL_CONFIG_DATA_PATH, "utf8");
const printConfigFile = await fs.readFile(PRINT_CONFIG_PATH, "utf8");

const normalConfigDataFieldsMatch = normalConfigDataFile.match(/type NormalConfigData struct {([^}]*)}/s);
if (!normalConfigDataFieldsMatch) {
  throw new Error("Failed to find NormalConfigData struct");
}
const normalConfigDataFields = normalConfigDataFieldsMatch[1].trim().split("\n").map((line) =>
  line.trim().split(/\s+/)[0]
).filter(field => !field.startsWith("//"));

const printConfigFunctionMatch = printConfigFile.match(/func printConfig\(.*?\) {([^}]*)}/s);
if (!printConfigFunctionMatch) {
  throw new Error("Failed to find printConfig function");
}
const printConfigFunctionBody = printConfigFunctionMatch[1];

const missingFields = normalConfigDataFields.filter((field) =>
  !printConfigFunctionBody.includes(field) && !ACCEPTABLE_MISSING_FIELDS.includes(field)
);
if (missingFields.length > 0) {
  console.error("Missing fields in printConfig function:");
  console.log(missingFields.join("\n"));
  process.exit(1);
}

File: tools/print_config_exhaustive/lint.go
package main

import (
	"fmt"
	"log"
	"os"
	"regexp"
	"slices"
	"strings"
)

const (
	normalConfigDataPath = "internal/config/normal_config.go"
	printConfigPath      = "internal/cmd/config/root.go"
)

var whiteList = []string{
	"Aliases",
	"BranchTypeOverrides",
	"DryRun",
	"ProposalsShowLineage", // TODO: remove once https://github.com/git-town/git-town/issues/3003 is shipped
	"Verbose",
}

func main() {
	definedFields := FindDefinedFields(readFile(normalConfigDataPath))
	printFunction := FindPrintFunc(readFile(printConfigPath))
	unprinted := FindUnprintedFields(definedFields, printFunction, whiteList)
	if len(unprinted) > 0 {
		printMissingFields(unprinted)
		os.Exit(1)
	}
}

func FindDefinedFields(text string) []string {
	structRE := regexp.MustCompile(`type NormalConfig struct {([^}]*)}`)
	match := structRE.FindStringSubmatch(text)
	if len(match) < 2 {
		log.Fatalf("Error: Failed to find NormalConfig struct")
	}
	result := []string{}
	for line := range strings.SplitSeq(strings.TrimSpace(match[1]), "\n") {
		trimmedLine := strings.TrimSpace(line)
		if strings.HasPrefix(trimmedLine, "//") {
			continue // skip comments
		}
		parts := strings.Fields(trimmedLine)
		if len(parts) > 0 {
			result = append(result, parts[0])
		}
	}
	return result
}

func FindPrintFunc(text string) string {
	funcBodyRE := regexp.MustCompile(`func printConfig\(.*?\) {([^}]*)}`)
	match := funcBodyRE.FindStringSubmatch(text)
	if len(match) < 2 {
		log.Fatalf("Error: Failed to find printConfig function")
	}
	return match[1]
}

func FindUnprintedFields(fields []string, text string, whiteList []string) []string {
	result := []string{}
	for _, field := range fields {
		if !isPrinted(field, text) && !isWhitelisted(field, whiteList) {
			result = append(result, field)
		}
	}
	return result
}

func isPrinted(field, text string) bool {
	return strings.Contains(text, field)
}

func isWhitelisted(field string, whitelist []string) bool {
	return slices.Contains(whitelist, field)
}

func printMissingFields(unprinted []string) {
	fmt.Println("Missing fields in printConfig function:")
	for _, field := range unprinted {
		fmt.Println(field)
	}
}

func readFile(path string) string {
	result, err := os.ReadFile(path)
	if err != nil {
		log.Fatalf("Error reading %s: %v\n", path, err)
	}
	return string(result)
}
